src/qt/nametablemodel.cpp

```diff
@@ -0,0 +1,283 @@
+#include <qt/nametablemodel.h>
+
+#include <chainparams.h>
+#include <interfaces/node.h>
+#include <names/common.h>
+#include <names/encoding.h>
+#include <names/main.h>
+#include <qt/clientmodel.h>
+#include <qt/guiconstants.h>
+#include <qt/guiutil.h>
+#include <qt/platformstyle.h>
+#include <qt/transactiontablemodel.h>
+#include <qt/walletmodel.h>
+#include <rpc/protocol.h>
+#include <rpc/server.h>
+#include <node/ui_interface.h>
+#include <univalue.h>
+#include <util/ref.h>
+#include <validation.h> // cs_main
+#include <wallet/wallet.h>
+
+#include <QDebug>
+#include <QObject>
+#include <QTimer>
+
+// ExpiresIn column is right-aligned as it contains numbers
+namespace {
+    int column_alignments[] = {
+        Qt::AlignLeft|Qt::AlignVCenter,     // Name
+        Qt::AlignLeft|Qt::AlignVCenter,     // Value
+        Qt::AlignRight|Qt::AlignVCenter,    // Expires in
+        Qt::AlignRight|Qt::AlignVCenter,    // Name Status
+    };
+}
+
+// Returns true if new height is better
+bool NameTableEntry::CompareHeight(int nOldHeight, int nNewHeight)
+{
+    if(nOldHeight == NAME_NON_EXISTING)
+        return true;
+
+    // We use optimistic way, assuming that unconfirmed transaction will eventually become confirmed,
+    // so we update the name in the table immediately. Ideally we need a separate way of displaying
+    // unconfirmed names (e.g. grayed out)
+    if(nNewHeight == NAME_UNCONFIRMED)
+        return true;
+
+    // Here we rely on the fact that dummy height values are always negative
+    return nNewHeight > nOldHeight;
+}
+
+// Private implementation
+class NameTablePriv
+{
+public:
+    explicit NameTablePriv(NameTableModel *_parent) :
```

Comment:
If `_parent` is never null (not sure, but the code sometimes just dereferences it without checking), then perhaps this could be made a reference instead.

Reply:
Indeed, it is never null.  I'll make it a reference.